### DISCUSSION TITLE
Websocket closing after receiving or sending a message

### DISCUSSION THREAD

**Discussion by auroraveon** [cc-tweaked/CC-Tweaked#988] [Q&A] [answered]
### Minecraft Version

1.16.x

### Version

1.99.1

### Details

I'm currently working on a project that uses a python websocket server to communicate to turtles running their respective client code.

However, after the lua websocket client receives or sends a message, it closes the websocket and then throws the following error after it tries to receive or send (in this case receive - see client code below):
```lua
main.lua:14: attempt to use a closed file
```

The client function loop (only the loop is included as the other stuff is just variables):
```lua
function client()
    local ws, err = http.websocket(URL) -- the websocket
    if err then
        print('Error: '..err)
    elseif ws then
        print('Connected! Listening on port '..tostring(PORT))

        while true do
            local msg = ws.receive() -- <--- this is where the error occurs from

            if msg == nil then
                break
            elseif string.find(msg, "^Server:") then
                print(msg)
            else
                local func , err = load(msg) -- code from the server
                if not err then
                    print('Executing code: '..msg)
                    local result = func () -- the return data
                    ws.send(json:encode({data=result}))
                else
                    print('Error: '..'Remote code could not execute.')
                end
            end
        end
    end
end
```

Here is another websocket client implementation that is much simpler and still has the same error:
```lua
local ws, err = http.websocket("ws://127.0.0.1:7890")

if err then
    print(err)
elseif ws then
    print('Connected!')
    while true do
        local msg = ws.receive()
        print(msg)
    end
end
```

And this gives this output to the turtle's console:
![image](https://user-images.githubusercontent.com/78045343/145900514-6485964c-5248-4935-b520-766fbdb4342e.png)

Also please excuse any bad lua code practices because I'm quite new to it.

I'm happy to provide any other information required to find the issue.

**Comment by SquidDev** [comment, marked as answer]
For now I'm going to assume this is a problem with your webserver's websocket code. Would you be able to post that too?

**Reply by auroraveon** [reply]
Yep, here is the code:

```py
# libraries for the websocket
import websockets
from websockets.server import WebSocketServerProtocol
import asyncio

'''
A Websocket server built for CC: Tweaked
Communicates to turtles inside of Minecraft
'''

PORT = 7890
URL = f'ws://127.0.0.1:{PORT}'

print(f'Starting server on port {PORT}...')

async def ws_loop(ws: WebSocketServerProtocol, path):
    print(f'Client connecting...')

    await ws.send('turtle.forward()')
    await ws.send('turtle.forward()')
    await ws.send('turtle.forward()')

    try:
        async for msg in ws:
            print(f"Received message from client: {msg}")
            await ws.send('turtle.forward()')
    except websockets.exceptions.ConnectionClosed as e:
        print('Client disconnected.')

server = websockets.serve(ws_loop, "localhost", PORT)

loop = asyncio.get_event_loop() # asyncio event loop

async def main():
    # start the server
    await server
    print('Server started.')

    print(f'Server successfully started on port {PORT}!')

asyncio.get_event_loop().run_until_complete(main())
asyncio.get_event_loop().run_forever()
```

**Reply by SquidDev** [reply]
I had issues getting CC:T to even connect to the server (Python 3.10, websockets 10.1), which looks like another instance of #958.

Adding `compression=None` to the `websockets.serve` argument did appear to fix that, after which I wasn't able to reproduce your issue. I guess it's worth a try at least?

**Reply by auroraveon** [reply]
I tried adding `compression=None` and now the client on my turtle is working.
I'm using the same websockets version as you however I'm using Python 3.9.7 which could be the problem.

I never read that issue that you referenced there prior to now, therefore didn't know that it currently doesn't support compression.

However, thank you so much for helping, I appreciate it.

Note:
While I was waiting for a reply after sending the code, I looked around at other discussions involving websockets and realised that others were using JavaScript for their websocket servers, so I decided to implement one too in Node.js. I don't know the exact differences between the websockets module in Python and the one in , but maybe due to the change of adding `compression=None` in the Python websockets server, it could suggest that Node.js's websocket module is by default, not using compression. Just a theory as I haven't looked into it yet.
